src/agent/coverage/src/cobertura.rs

```diff
@@ -0,0 +1,245 @@
+use crate::source::SourceCoverage;
+use crate::source::SourceCoverageLocation;
+use crate::source::SourceFileCoverage;
+use anyhow::Context;
+use anyhow::Error;
+use anyhow::Result;
+use std::time::{SystemTime, UNIX_EPOCH};
+use xml::writer::{EmitterConfig, XmlEvent};
+
+pub fn cobertura(source_coverage: SourceCoverage) -> Result<String, Error> {
+    let mut backing: Vec<u8> = Vec::new();
+    let mut emitter = EmitterConfig::new()
+        .perform_indent(true)
+        .create_writer(&mut backing);
+
+    let unixtime = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .context("system time before unix epoch")?
+        .as_secs();
+
+    emitter.write(
+        XmlEvent::start_element("coverage")
+            .attr("lines-valid", "0")
+            .attr("lines-covered", "0")
+            .attr("line-rate", "0")
+            .attr("branches-valid", "0")
+            .attr("branches-covered", "0")
+            .attr("branch-rate", "0")
+            .attr("timestamp", &format!("{}", unixtime))
+            .attr("complexity", "0")
+            .attr("version", "0.1"),
+    )?;
+    emitter.write(XmlEvent::start_element("sources"))?;
+    emitter.write(XmlEvent::start_element("source"))?;
+    emitter.write(XmlEvent::characters(""))?;
+    emitter.write(XmlEvent::end_element())?; // source
+    emitter.write(XmlEvent::end_element())?; // sources
+
+    emitter.write(XmlEvent::start_element("packages"))?;
+    emitter.write(
+        XmlEvent::start_element("package")
+            .attr("name", "0")
+            .attr("lines-valid", "0")
+            .attr("lines-covered", "0")
+            .attr("line-rate", "0")
+            .attr("branches-valid", "0")
+            .attr("branches-covered", "0")
+            .attr("branch-rate", "0")
+            .attr("complexity", "0"),
+    )?;
+    emitter.write(XmlEvent::start_element("classes"))?;
+    // loop through files
+    let files: Vec<SourceFileCoverage> = source_coverage.files;
+    for file in files {
+        emitter.write(
+            XmlEvent::start_element("class")
+                .attr("name", "0")
+                .attr("filename", &file.file)
+                .attr("lines-valid", "0")
+                .attr("lines-covered", "0")
+                .attr("line-rate", "0")
+                .attr("branches-valid", "0")
+                .attr("branches-covered", "0")
+                .attr("branch-rate", "0")
+                .attr("complexity", "0"),
+        )?;
+        // emitter.write(XmlEvent::start_element("methods"))?;
```

Comment:
I removed methods from the XML output for now because it included things source format doesn't currently have such as signature.

Reply:
Based on [this schema](https://github.com/danielpalme/ReportGenerator), I think zero (or more) elements are valid for these fields we don't have real data for.

What matters most, though, is how [ReportGenerator](https://github.com/danielpalme/ReportGenerator) renders the XML as HTML. This is because it is what [powers the ADO task](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PublishCodeCoverageResultsV2/ThirdPartyNotices.txt#L7).

Let's investigate what happens in ADO.